### PR TITLE
MARKET-1808 Particles component

### PR DESCRIPTION
**Pull request by serhmel**
[UI-Builder](https://develop.backendless.com/app/serhiy_melnychuk_components/ui-builder/bl-particles/pages/landing)
[Desktop Preview](https://genteelinstrument.backendless.app/api/files/ui-builder/containers/bl-particles/index.html?page=landing)

big-circles
growing
light
fireworks
fire

UI Builder Preview:
ui-builder-preview

### CODE REVIEW

File: components/bl-particles-component/README.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+| Number <br> `number`                                     | *Number*                                                                                                                                                                                                                                | 15            |                      | NO           | YES        | Controls the number of particles displayed on the canvas.                                                                                                                                           |
+| Size <br> `size`                                         | *Number*                                                                                                                                                                                                                                | 5             | Size Logic           | NO           | YES        | Controls the size of particles displayed on the canvas. Specify an object with minimum and maximum values by Size Logic to use a range for the size. Watch [Codeless Examples](#codeless-examples). |
```
[the remark this answers is not in the document]
Comment (serhmel) [reply components/bl-particles-component/README.md:26]: The links will work after the component is merged.

File: components/bl-particles-component/src/index.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+function useComponentActions(component, containerRef) {
+  Object.assign(component, {
+    start  : () => containerRef.current?.start(),
+    stop   : () => containerRef.current?.stop(),
+    play   : () => containerRef.current?.play(),
+    pause  : () => containerRef.current?.pause(),
+    destroy: () => containerRef.current?.destroy(),
+    refresh: () => containerRef.current?.refresh(),
+  });
```
Comment (v-excelsior) [inline components/bl-particles-component/src/index.js:51]:
```
if(containerRef.current) {
 Object.assign(..)
}
```

?

File: components/bl-particles-component/src/helpers.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+export async function loadParticles(engine) {
+  await loadFull(engine);
+  await loadPlugins(engine);
+  await loadInteractions(engine);
+  await loadUpdaters(engine);
+  await loadPaths(engine);
+  await loadShapes(engine);
+}
```
Comment (v-excelsior) [inline components/bl-particles-component/src/helpers.js:191]: Do we need to do it synchronously?
`Promise.all` if no.

Also, all of this functions needed for every particles type?

File: components/bl-particles-component/src/helpers.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  await loadSimplexNoisePath(engine);
+}
+
+async function loadInteractions(engine) {
+  await loadLightInteraction(engine);
+  await loadParticlesRepulseInteraction(engine);
+}
+
+export async function loadParticles(engine) {
+  await loadFull(engine);
```
Comment (v-excelsior) [inline components/bl-particles-component/src/helpers.js:185]: image

Do we need to load everything else if we use `loadFull`?

File: components/bl-particles-component/src/index.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+export default function ParticlesComponent({ component, eventHandlers: { onLoad }, elRef, instanceId }) {
+  const { classList, display, style, width, height } = component;
+
+  const containerRef = useRef(null);
+
+  const options = useOptions(component);
+
+  const onInit = useCallback(async engine => await loadParticles(engine), []);
+  const onLoaded = useCallback(async ({ sourceOptions, canvas }) => {
```
Comment (v-excelsior) [inline components/bl-particles-component/src/index.js:16]: async?

File: components/bl-particles-component/src/helpers.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    rollingAnimation, duration, delay, triangleLinksVisibility, collisionMode, outModes, backgroundOpacity,
+  } = component;
+
+  const particles = useMemo(() => ({
+    collisions: { enable: collisionMode !== UNSET_COLLISION_MODE, mode: collisionMode },
+    links     : {
+      enable   : lineLinksVisibility || triangleLinksVisibility,
+      distance : LINKS_DISTANCE,
+      color    : linksColor || Colors.RANDOM,
+      opacity  : OPACITY[lineLinksVisibility ? 'FULL' : 'LOW'],
```
Comment (v-excelsior) [inline components/bl-particles-component/src/helpers.js:89]: `lineLinksVisibility ? OPACITY.FULL : OPACITY.LOW`

same for others

We use this maps for autocomplete, your way for using it miss the sense of it.

Also `OPACITY` -> `Opacity`

File: components/bl-particles-component/src/helpers.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    delay       : delay || DEFAULT_TIMING,
+    fullScreen  : { enable: fullScreen, zIndex: zIndex || DEFAULT_Z_INDEX },
+    background  : { color: backgroundColor, opacity: backgroundOpacity },
+    detectRetina: true,
+    pauseOnBlur : false,
+    ...PresetsMap[preset],
+    ...customOptions,
+  }), [
+    autoPlay, particles, duration, delay, fullScreen, zIndex, backgroundColor, backgroundOpacity, preset, customOptions,
+  ]);
```
Comment (v-excelsior) [inline components/bl-particles-component/src/helpers.js:137]: This useMemo shouldn't depend on `particles`, `autoPlay` and `customOptions` as i see

```suggestion
  const otherProps = useMemo(() => ({
    duration    : duration || DEFAULT_TIMING,
    delay       : delay || DEFAULT_TIMING,
    fullScreen  : { enable: fullScreen, zIndex: zIndex || DEFAULT_Z_INDEX },
    background  : { color: backgroundColor, opacity: backgroundOpacity },
    ...PresetsMap[preset],
  }), [ duration, delay, fullScreen, zIndex, backgroundColor, backgroundOpacity, preset  ]);
  
  
  return {
    particles, autoPlay,
     ...otherProps,    ...customOptions, 
     detectRetina: true,    pauseOnBlur : false 
   }
```

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-particles-component/src/index.js (hunk 2 above)
Comment (serhmel) [reply components/bl-particles-component/src/index.js:51]: Not sure it's a good idea.
In this case, the actions will not work if we set the display property to false for the first component rendering.

File: components/bl-particles-component/src/helpers.js (hunk 4 above)
Comment (serhmel) [reply components/bl-particles-component/src/helpers.js:185]:
```
export async function loadFull(engine: Engine): Promise<void> {
    await loadSlim(engine);

    await loadDestroyUpdater(engine);
    await loadRollUpdater(engine);
    await loadTiltUpdater(engine);
    await loadTwinkleUpdater(engine);
    await loadWobbleUpdater(engine);

    await loadExternalTrailInteraction(engine);

    await loadAbsorbersPlugin(engine);
    await loadEmittersPlugin(engine);
}
```

I added only those plugins that are not included in the loadFull function.

File: components/bl-particles-component/src/helpers.js (hunk 3 above)
Comment (serhmel) [reply components/bl-particles-component/src/helpers.js:191]: Not all of these functions are required for each particle type, but `onInit` with `loadParticles` runs only once on the first render, so we have to consider all possible configurations.

**Pull request merged**